File: packages-js/kanban-core/src/boardCommands.ts
/** Reine Board-Befehle: anlegen, umbenennen/anheften, Spalten konfigurieren, teilen, widerrufen. */
import { bump, requireAction, type Change, type CommandContext, type CommandResult } from './commands'
import { KanbanError, raiseIfDenied } from './errors'
import { cardsIn, DEFAULT_COLUMNS, findColumn, SCHEMA_VERSION, shareFor, type Board, type Card, type Column, type SharePermission, type TransitionPolicy } from './model'
import { checkRevoke, checkShare } from './permissions'
import { rankBetween, spreadRanks } from './rank'
import type { BoardTemplate } from './templates'
import { DEFAULT_LIMITS, validateBoardTitle, validateColumns } from './validation'

function result(board: Board, change: Change): CommandResult {
  return { board, changes: [change], card: null, warnings: [] }
}

export function createBoard(ctx: CommandContext, boardId: string, title = 'Neues Board', icon = '📋', template?: BoardTemplate): CommandResult {
  const limits = ctx.limits ?? DEFAULT_LIMITS
  const board: Board = {
    schema_version: SCHEMA_VERSION, id: boardId, title: validateBoardTitle(title, limits), icon, owner_id: ctx.actor,
    version: 1, pinned: false, archived: false, created_at: ctx.now, updated_at: ctx.now,
    columns: template ? validateColumns(template.columns, limits) : DEFAULT_COLUMNS.map((column) => ({ ...column })),
    cards: [], labels: [], shares: [],
    transitions: template?.transitions ?? { mode: 'free', allowed: [], locked_columns: [], fixed_order_columns: [] },
    wip_mode: 'block', extra: {},
  }
  return result(board, { kind: 'board.created', card_id: null, data: { template: template?.key ?? null } })
}

export interface BoardPatch {
  title?: string
  icon?: string
  pinned?: boolean
  archived?: boolean
}

/** Umbenennen (Eigentümer, Bearbeiten), Anheften/Archivieren (nur Eigentümer). */
export function updateBoard(board: Board, ctx: CommandContext, patch: BoardPatch): CommandResult {
  let next = board
  const changed: string[] = []
  if (patch.title !== undefined || patch.icon !== undefined) {
    requireAction(board, ctx, 'rename')
    const title = patch.title === undefined ? board.title : validateBoardTitle(patch.title, ctx.limits ?? DEFAULT_LIMITS)
    next = { ...next, title, icon: patch.icon ?? board.icon }
    changed.push(...(['title', 'icon'] as const).filter((name) => patch[name] !== undefined))
  }
  if (patch.pinned !== undefined || patch.archived !== undefined) {
    requireAction(board, ctx, 'pin')
    next = { ...next, pinned: patch.pinned ?? board.pinned, archived: patch.archived ?? board.archived }
    changed.push(...(['pinned', 'archived'] as const).filter((name) => patch[name] !== undefined))
  }
  return result(bump(next, ctx), { kind: 'board.updated', card_id: null, data: { fields: changed.sort() } })
}

function restrictPolicy(policy: TransitionPolicy, ids: ReadonlySet<string>): TransitionPolicy {
  return {
    mode: policy.mode,
    allowed: policy.mode === 'restricted' ? policy.allowed.filter(([from, to]) => ids.has(from) && ids.has(to)) : [],
    locked_columns: policy.locked_columns.filter((id) => ids.has(id)).sort(),
    fixed_order_columns: policy.fixed_order_columns.filter((id) => ids.has(id)).sort(),
  }
}

/** Karten entfernter Spalten gehen in bisheriger Reihenfolge ans Ende der Zielspalte. */
function rehome(board: Board, removed: ReadonlySet<string>, target: string): { cards: Card[]; moved: string[] } {
  const orphans = board.columns.filter((column) => removed.has(column.id)).flatMap((column) => cardsIn(board, column.id))
  if (orphans.length === 0) return { cards: board.cards, moved: [] }
  const last = cardsIn(board, target)
  let low = last[last.length - 1]?.rank ?? null
  const replacement = new Map<string, Card>()
  for (const orphan of orphans) {
    low = rankBetween(low, null)
    replacement.set(orphan.id, { ...orphan, column_id: target, rank: low })
  }
  return { cards: board.cards.map((card) => replacement.get(card.id) ?? card), moved: orphans.map((card) => card.id) }
}

/** Ersetzt den Spaltensatz (nur Eigentümer); Karten entfernter Spalten wandern in die erste Spalte. */
export function configureColumns(board: Board, ctx: CommandContext, columns: readonly Column[], transitions?: TransitionPolicy): CommandResult {
  requireAction(board, ctx, 'configure')
  const checked = validateColumns(columns, ctx.limits ?? DEFAULT_LIMITS)
  const ids = new Set(checked.map((column) => column.id))
  const removed = new Set(board.columns.map((column) => column.id).filter((id) => !ids.has(id)))
  const staged: Board = { ...board, columns: checked, transitions: restrictPolicy(transitions ?? board.transitions, ids) }
  const { cards, moved } = rehome(board, removed, (checked[0] as Column).id)
  const data = { removed: [...removed].sort(), moved_cards: moved }
  return result(bump(staged, ctx, cards), { kind: 'board.configured', card_id: null, data })
}

/** Freigabe anlegen oder ändern (Upsert); nur der Eigentümer darf teilen. */
export function shareBoard(board: Board, ctx: CommandContext, userId: string, permission: string): CommandResult {
  raiseIfDenied(checkShare(board, ctx.actor, userId, permission))
  const granted = permission as SharePermission
  const existing = shareFor(board, userId)
  const shares = existing
    ? board.shares.map((share) => (share.user_id === userId ? { ...share, permission: granted } : share))
    : [...board.shares, { user_id: userId, permission: granted, shared_by: ctx.actor, created_at: ctx.now }]
  const kind = existing ? 'share.updated' : 'share.created'
  return result(bump({ ...board, shares }, ctx), { kind, card_id: null, data: { user_id: userId, permission } })
}

/** Eigentümer widerruft jede Freigabe, Empfänger die eigene. */
export function revokeShare(board: Board, ctx: CommandContext, userId: string): CommandResult {
  raiseIfDenied(checkRevoke(board, ctx.actor, userId))
  const shares = board.shares.filter((share) => share.user_id !== userId)
  return result(bump({ ...board, shares }, ctx), { kind: 'share.revoked', card_id: null, data: { user_id: userId } })
}

/** Ersetzt die Ränge einer Spalte durch gleichmäßig verteilte Schlüssel (Reihenfolge bleibt). */
export function respreadColumn(board: Board, ctx: CommandContext, columnId: string): CommandResult {
  requireAction(board, ctx, 'move_card')
  if (!findColumn(board, columnId)) throw new KanbanError('UNKNOWN_COLUMN', `Unbekannte Spalte '${columnId}'`)
  const ordered = cardsIn(board, columnId)
  const keys = spreadRanks(ordered.length)
  const ranks = new Map(ordered.map((card, index) => [card.id, keys[index] as string]))
  const cards = board.cards.map((card) => ({ ...card, rank: ranks.get(card.id) ?? card.rank }))
  return result(bump(board, ctx, cards), { kind: 'column.rebalanced', card_id: null, data: { ranks: Object.fromEntries(ranks) } })
}

File: packages/auditcore_kanban/src/auditcore_kanban/board_commands.py
"""Pure board-level commands: create, rename/pin, configure columns, share, revoke."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import replace

from .commands import CommandResult, Context, bump, require
from .errors import JsonObject, KanbanError
from .events import Change
from .model import Board, Card, Column, Share, TransitionPolicy
from .permissions import Action, check_revoke, check_share
from .rank import rank_between, spread_ranks
from .templates import BoardTemplate
from .validation import validate_board_title, validate_columns


def create_board(
    ctx: Context,
    board_id: str,
    title: str = "Neues Board",
    icon: str = "📋",
    board_template: BoardTemplate | None = None,
) -> CommandResult:
    """New board owned by the acting user (default columns unless a template is given)."""
    board = Board(
        id=board_id,
        title=validate_board_title(title, ctx.limits),
        owner_id=ctx.actor,
        icon=icon,
        version=1,
        created_at=ctx.now,
        updated_at=ctx.now,
    )
    if board_template is not None:
        board = replace(
            board,
            columns=validate_columns(board_template.columns, ctx.limits),
            transitions=board_template.transitions,
        )
    data: JsonObject = {"template": board_template.key if board_template else None}
    return CommandResult(board, (Change("board.created", None, data),))


def update_board(
    board: Board,
    ctx: Context,
    *,
    title: str | None = None,
    icon: str | None = None,
    pinned: bool | None = None,
    archived: bool | None = None,
) -> CommandResult:
    """Rename (owner, edit), pin/archive (owner only)."""
    changed: list[str] = []
    if title is not None or icon is not None:
        require(board, ctx, Action.RENAME)
        board = replace(
            board,
            title=board.title if title is None else validate_board_title(title, ctx.limits),
            icon=board.icon if icon is None else icon,
        )
        changed += [n for n, v in (("title", title), ("icon", icon)) if v is not None]
    if pinned is not None or archived is not None:
        require(board, ctx, Action.PIN)
        board = replace(
            board,
            pinned=board.pinned if pinned is None else pinned,
            archived=board.archived if archived is None else archived,
        )
        changed += [n for n, v in (("pinned", pinned), ("archived", archived)) if v is not None]
    names = sorted(changed)
    data: JsonObject = {"fields": [n for n in names]}
    return CommandResult(bump(board, ctx), (Change("board.updated", None, data),))


def _restricted_policy(policy: TransitionPolicy, ids: set[str]) -> TransitionPolicy:
    allowed = None
    if policy.allowed is not None:
        allowed = frozenset(p for p in policy.allowed if p[0] in ids and p[1] in ids)
    return replace(
        policy,
        allowed=allowed,
        locked_columns=frozenset(policy.locked_columns & ids),
        fixed_order_columns=frozenset(policy.fixed_order_columns & ids),
    )


def _rehome(board: Board, removed: set[str], target: str) -> tuple[tuple[Card, ...], list[str]]:
    """Cards of removed columns go to the end of ``target`` in their previous order."""
    orphans = [c for col in board.columns if col.id in removed for c in board.cards_in(col.id)]
    if not orphans:
        return board.cards, []
    last = board.cards_in(target)
    low = last[-1].rank if last else None
    ranks: list[str] = []
    for _ in orphans:
        low = rank_between(low, None)
        ranks.append(low)
    new = {c.id: replace(c, column_id=target, rank=r) for c, r in zip(orphans, ranks, strict=True)}
    return tuple(new.get(c.id, c) for c in board.cards), [c.id for c in orphans]


def configure_columns(
    board: Board,
    ctx: Context,
    columns: Sequence[Column],
    transitions: TransitionPolicy | None = None,
) -> CommandResult:
    """Replace the column set (owner only); cards of removed columns move to the first column."""
    require(board, ctx, Action.CONFIGURE)
    checked = validate_columns(columns, ctx.limits)
    ids = {c.id for c in checked}
    removed = {c.id for c in board.columns} - ids
    policy = _restricted_policy(transitions or board.transitions, ids)
    staged = replace(board, columns=checked, transitions=policy)
    cards, moved = _rehome(board, removed, checked[0].id)
    data: JsonObject = {"removed": [c for c in sorted(removed)], "moved_cards": [c for c in moved]}
    return CommandResult(bump(staged, ctx, cards), (Change("board.configured", None, data),))


def share_board(
    board: Board, ctx: Context, user_id: str, permission: str
) -> CommandResult:
    """Grant or update (upsert) a share; the owner is the only one who may share."""
    check_share(board, ctx.actor, user_id, permission).raise_if_denied()
    existing = board.share_for(user_id)
    if existing is not None:
        shares = tuple(
            replace(s, permission=permission) if s.user_id == user_id else s for s in board.shares
        )
        kind = "share.updated"
    else:
        shares = board.shares + (Share(user_id, permission, ctx.actor, ctx.now),)
        kind = "share.created"
    data: JsonObject = {"user_id": user_id, "permission": permission}
    return CommandResult(bump(replace(board, shares=shares), ctx), (Change(kind, None, data),))


def revoke_share(board: Board, ctx: Context, user_id: str) -> CommandResult:
    """Owner revokes any share, a recipient their own."""
    check_revoke(board, ctx.actor, user_id).raise_if_denied()
    shares = tuple(s for s in board.shares if s.user_id != user_id)
    data: JsonObject = {"user_id": user_id}
    return CommandResult(
        bump(replace(board, shares=shares), ctx), (Change("share.revoked", None, data),)
    )


def respread_column(board: Board, ctx: Context, column_id: str) -> CommandResult:
    """Replace the ranks of one column by evenly spaced keys (keeps the order)."""
    require(board, ctx, Action.MOVE_CARD)
    if board.column(column_id) is None:
        raise KanbanError("UNKNOWN_COLUMN", f"Unbekannte Spalte '{column_id}'")
    ordered = board.cards_in(column_id)
    ranks = dict(zip((c.id for c in ordered), spread_ranks(len(ordered)), strict=True))
    cards = tuple(replace(c, rank=ranks[c.id]) if c.id in ranks else c for c in board.cards)
    data: JsonObject = {"ranks": {k: v for k, v in ranks.items()}}
    return CommandResult(bump(board, ctx, cards), (Change("column.rebalanced", None, data),))
